pennylane/data/dataset.py

```diff
@@ -240,52 +245,57 @@ def __copy__(self):
+        copied._attr_filemap = self._attr_filemap.copy()
         copied.__doc__ = self.__doc__
         for key, val in self.attrs.items():
             setattr(copied, f"{key}", val)
 
         return copied
 
-    # The methods below are intended for use only by standard Dataset objects
-    def __get_filename_for_attribute(self, attribute):
-        return self._fullfile or self._prefix.format(attribute)
-
-    def __get_attribute_from_filename(self, filename):
-        return os.path.basename(filename)[self._prefix_len : -4]
-
     def __getattribute__(self, name):
-        try:
-            val = super().__getattribute__(name)
-            if val is None and self._is_standard and name in self.attrs:
-                raise AttributeError(
-                    f"Dataset has a '{name}' attribute, but it is None and no data file was found"
-                )
-            return val
-        except AttributeError:
-            if not self._is_standard:
-                raise
-            filepath = self.__get_filename_for_attribute(name)
-            if not os.path.exists(filepath):  # TODO: download the file here?
-                raise
-            value = self._read_file(filepath)
-            if filepath == self._fullfile:
-                if name not in value:
-                    raise
-                value = value[name]
-            if name in condensed_hamiltonians:
-                value = dict_to_hamiltonian(value["terms"], value["wire_map"])
-            setattr(self, name, value)
+        value = super().__getattribute__(name)
+        if value is not None or name not in self.attrs:
             return value
 
+        filepath = self._attr_filemap.get(name)
+        if not filepath:  # this is (hopefully) an attribute that was directly set to None
+            return None
+        if not os.path.exists(filepath):
+            raise DatasetLoadError(
+                f"Dataset lazy-loaded a '{name}' attribute, but the file originally containing it ({filepath}) was not found."
+            )
+
+        value = self.get_file_contents(filepath)
+        # if this is a standard dataset getting the attribute from a special single-attribute
+        # file, then `value` is already set to the value of the attribute
+        if not (
+            self._is_standard
+            and os.path.exists(self._prefix.format(name))
+            and os.path.samefile(filepath, self._prefix.format(name))
+        ):
+            if name not in value:
+                raise DatasetLoadError(
+                    f"Dataset lazy-loaded a '{name}' attribute from {filepath}, but it no longer appears to be in the file."
+                )
+            value = value[name]
+        if name in condensed_hamiltonians:
+            value = dict_to_hamiltonian(value["terms"], value["wire_map"])
+
+        setattr(self, name, value)
+        del self._attr_filemap[name]
+        return value
+
 
 def dict_to_hamiltonian(terms, wire_map):
```

Comment:
Can we add examples for both conversion functions? To make things a bit more explicit for users who come to find them.

Reply:
now that you mention it, I'm not sure I want these methods to be so user-facing (which is why I didn't import them in the `__init__` file). They shouldn't use this ever, and I am going to make them private again.